crates/vim/src/surrounds.rs

```diff
@@ -0,0 +1,589 @@
+use crate::{motion::Motion, object::Object, state::Mode, Vim};
+use editor::{scroll::Autoscroll, Bias};
+use gpui::WindowContext;
+use language::BracketPair;
+use serde::Deserialize;
+use std::ops::Deref;
+use std::sync::Arc;
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize)]
+pub enum SurroundsType {
+    Motion(Motion),
+    Object(Object),
+}
+
+pub fn add_surrounds(text: Arc<str>, target: SurroundsType, cx: &mut WindowContext) {
+    Vim::update(cx, |vim, cx| {
+        vim.stop_recording();
+        vim.update_active_editor(cx, |_, editor, cx| {
+            let text_layout_details = editor.text_layout_details(cx);
+            editor.transact(cx, |editor, cx| {
+                editor.set_clip_at_line_ends(false, cx);
+                editor.change_selections(Some(Autoscroll::fit()), cx, |s| {
+                    s.move_with(|map, selection| match &target {
+                        SurroundsType::Object(object) => {
+                            object.expand_selection(map, selection, false);
+                        }
+                        SurroundsType::Motion(motion) => {
+                            motion.expand_selection(
+                                map,
+                                selection,
+                                Some(1),
+                                true,
+                                &text_layout_details,
+                            );
+                        }
+                    });
+                });
+
+                let input_text = text.to_string();
+                let pair = match find_surround_pair(&all_support_surround_pair(), text.deref()) {
+                    Some(pair) => pair,
+                    None => return,
+                };
```

Comment:
@weartist In the case that no pair matches, we should use the typed character as open and close (like surround.vim).